# Metabolic Adaptation during Nutrient Scarcity Educates Multinuclear Giant Cell Formation

## Study authors

J.S. Brunner, L. Vulliard, M. Hofmann, M. Kieler, A. Lercher, A. Vogel, M. Kerndl, V. Saferding, B. Niederreiter, A. Frauenstein, C. Scherbel, Y. Mikami, G. Krönke, A. Bergthaler, J.J. O’Shea, T. Weichhart, F. Meissner, J. S. Smolen, P. Cheng, O. Yanes, J. Menche, P. Murray, O. Sharif, S. Blüml* and G. Schabbauer*.

Correspondence to: gernot.schabbauer(at)meduniwien.ac.at and stephan.blueml(at)meduniwien.ac.at.

## Data availability

Proteomics and transcriptomics data were deposited on public repositories and are necessary to reproduce the analyses.  
The RNA-seq data (BAM alignments and feature counts) have been deposited in NCBI's Gene Expression Omnibus (Edgar et al., 2002) and are accessible through [GEO Series accession number GSE125101](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE125101).

## Code summary

Several *R* Jupyter notebooks are provided and were used to perform the analyses shown in the corresponding paper.  
**All absolute paths used here and in the notebooks must be adapted to your own configuration.**

### RNA-seq analysis

Counts were generated using the *countExons.R* script. The corresponding feature counts are available from the GEO repository as well.  
See *RNA-DiffEq.ipynb*.

### Proteomics analysis

As described in the methods section of the article, the MS data were processed in Perseus (filtered for coefficients of variation bigger across conditions than between replicates, converted to z-score and compared with t-test between conditions of interest).  
See *MS-analysis.ipynb* for downstream analysis.

### Integration

See *DifferentialIntegration.ipynb*.  
The transcriptomics and proteomics results were integrated using the OmicsIntegrator (v0.3.1) with the following parameters:

	python2.7 /Users/lvulliard/bin/OmicsIntegrator-0.3.1/scripts/garnet.py --outdir="recArg1_output" recArg1_garnet.cfg
	python2.7 /Users/lvulliard/bin/OmicsIntegrator-0.3.1/scripts/forest.py --prize=recArg1_prot.tsv --edge=mousePPI.tsv --conf=forest.cfg --outpath=recArg1_output --msgpath=/Users/lvulliard/bin/msgsteiner-1.3/msgsteiner --garnet=/Users/lvulliard/OneShotProject/RecArg1/OmicsIntegratorRun/recArg1_output/events_to_genes_with_motifsregression_results_FOREST_INPUT.tsv --noisyEdges 50

	python2.7 /Users/lvulliard/bin/OmicsIntegrator-0.3.1/scripts/garnet.py --outdir="ranklDiff_output" ranklDiff_garnet.cfg
	python2.7 /Users/lvulliard/bin/OmicsIntegrator-0.3.1/scripts/forest.py --prize=ranklDiff_prot.tsv --edge=mousePPI.tsv --conf=forest.cfg --outpath=ranklDiff_output --msgpath=/Users/lvulliard/bin/msgsteiner-1.3/msgsteiner --garnet=ranklDiff_output/events_to_genes_with_motifsregression_results_FOREST_INPUT.tsv --noisyEdges 50

 The configuration files for Garnet and Forest (suffix in *.cfg*) are provided with this repository. The other files are generated using the notebooks provided, in which the corresponding depencies are detailed.  
 NB: the *events_to_genes_with_motifsregression_results_FOREST_INPUT.tsv* files generated by Garnet need convertion to accepted mouse gene symbols before running Forest, or the effect of the transcriptomics changes are greatly neglected. 

The *Python* notebook *InfomapNetwork.ipynb* was used to detect communities on the integrated networks using the version 1.0.0b15 of the Python package. Then, functional enrichments in each cluster were tested in the *InfomapAnalysis.ipynb R* notebook.



[![DOI](https://zenodo.org/badge/164895854.svg)](https://zenodo.org/badge/latestdoi/164895854)
